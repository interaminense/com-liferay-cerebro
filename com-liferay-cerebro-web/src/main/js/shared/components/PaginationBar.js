import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import getCN from 'classnames';
import {noop} from 'lodash';

import Dropdown from './Dropdown';
import NavBar from './NavBar';
import analyticsConstants from 'shared/util/constants';
import Pagination from './Pagination';
import {setUriQueryValue} from 'shared/util/router';
import {sub} from '../util/lang';

const {deltaValues} = analyticsConstants.pagination;

const SIZES = ['sm', 'lg'];

class DeltaItem extends Component {
	@autobind
	handleChange() {
		const {delta, onChange} = this.props;

		onChange(delta);
	}

	render() {
		return (
			<Dropdown.Item {...this.otherProps()} onClick={this.handleChange}>
				{this.props.children}
			</Dropdown.Item>
		);
	}
}

DeltaItem.PROPS = {
	delta: Config.number().required(),
	onChange: Config.func().value(noop)
};

class PaginationBar extends Component {
	render() {
		const {
			deltas,
			href,
			onDeltaChange,
			onPageChange,
			page,
			selectedDelta,
			size,
			totalItems
		} = this.props;

		const classes = getCN('pagination-bar-root', {
			[`pagination-${size}`]: size
		});

		const start = page * selectedDelta;

		return (
			<NavBar elementClasses={classes}>
				<Dropdown
					align="bottomLeft"
					buttonProps={{
						size: 'sm'
					}}
					elementClasses="pagination-items-per-page"
					label={selectedDelta}
				>
					{deltas.map(item => (
						<DeltaItem
							delta={item}
							href={
								onDeltaChange
									? null
									: setUriQueryValue(href, 'delta', item)
							}
							key={item}
							onChange={onDeltaChange}
						>
							{item}
						</DeltaItem>
					))}
				</Dropdown>

				<div class="pagination-results">
					{sub(Liferay.Language.get('showing-x-to-x-of-x-entries'), [
						start - selectedDelta + 1,
						Math.min(start, totalItems),
						totalItems
					])}
				</div>

				<Pagination
					href={href}
					onChange={onPageChange}
					page={page}
					total={Math.ceil(totalItems / selectedDelta)}
				/>
			</NavBar>
		);
	}
}

PaginationBar.PROPS = {
	deltas: Config.array().value(deltaValues),
	href: Config.string(),
	onDeltaChange: Config.func(),
	onPageChange: Config.func(),
	page: Config.number(),
	selectedDelta: Config.number().value(deltaValues[1]),
	size: Config.oneOf(SIZES),
	totalItems: Config.number()
};

PaginationBar.STATE = {};

export default PaginationBar;