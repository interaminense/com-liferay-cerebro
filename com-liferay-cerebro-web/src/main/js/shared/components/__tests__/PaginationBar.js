import PaginationBar from '../PaginationBar';
import {renderWithStore} from 'test/mock-store';

describe('PaginationBar', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = renderWithStore(PaginationBar, {
			href: '',
			page: 3,
			selectedDelta: 10,
			totalItems: 100
		});

		expect(component).toMatchSnapshot();
	});

	it('should render with buttons if passed an onPageChange handler', () => {
		component = renderWithStore(PaginationBar, {
			onPageChange: jest.fn(),
			page: 3,
			selectedDelta: 10,
			totalItems: 100
		});

		expect(component).toMatchSnapshot();
	});

	it('should render with small size', () => {
		component = renderWithStore(PaginationBar, {
			href: '',
			page: 3,
			selectedDelta: 10,
			size: 'sm',
			totalItems: 100
		});

		expect(component).toMatchSnapshot();
	});

	it('should render with large size', () => {
		component = renderWithStore(PaginationBar, {
			href: '',
			page: 3,
			selectedDelta: 10,
			size: 'lg',
			totalItems: 100
		});

		expect(component).toMatchSnapshot();
	});

	it('should render different deltas', () => {
		component = renderWithStore(PaginationBar, {
			deltas: [1, 2, 3, 4],
			href: '',
			page: 1,
			selectedDelta: 1,
			totalItems: 10
		});

		expect(component).toMatchSnapshot();
	});
});