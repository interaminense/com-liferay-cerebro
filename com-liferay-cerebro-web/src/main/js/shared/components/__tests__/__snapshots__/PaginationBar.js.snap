// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationBar should render 1`] = `
<nav class="navbar pagination-bar-root">
  <div class="dropdown dropdown-root pagination-items-per-page" id="overlay1">
    <button class="button-root btn btn-secondary btn-sm dropdown-toggle" type="button">
      <span class="text-truncate">10 </span>
      <span>
        <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
        </svg>
      </span>
    </button>
  </div>
  <div class="pagination-results">showing-x-to-x-of-x-entries</div>
  <ul class="pagination pagination-root">
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
          <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
        </svg>
      </button>
    </li>
    <li class="page-item">
      <button class="page-link">1</button>
    </li>
    <li class="page-item">
      <button class="page-link">2</button>
    </li>
    <li class="page-item active">
      <button class="page-link">3</button>
    </li>
    <li class="page-item">
      <button class="page-link">4</button>
    </li>
    <li class="page-item">
      <button class="page-link">5</button>
    </li>
    <li class="page-item">
      <div class="dropdown dropdown-root" id="overlay2">
        <button class="button-root btn btn-link nav-link dropdown-toggle page-link" type="button">
          <span class="text-truncate">... </span>
        </button>
      </div>
    </li>
    <li class="page-item">
      <button class="page-link">10</button>
    </li>
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
          <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
        </svg>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`PaginationBar should render different deltas 1`] = `
<nav class="navbar pagination-bar-root">
  <div class="dropdown dropdown-root pagination-items-per-page" id="overlay9">
    <button class="button-root btn btn-secondary btn-sm dropdown-toggle" type="button">
      <span class="text-truncate">1 </span>
      <span>
        <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
        </svg>
      </span>
    </button>
  </div>
  <div class="pagination-results">showing-x-to-x-of-x-entries</div>
  <ul class="pagination pagination-root">
    <li class="page-item disabled">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
          <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
        </svg>
      </button>
    </li>
    <li class="page-item active">
      <button class="page-link">1</button>
    </li>
    <li class="page-item">
      <button class="page-link">2</button>
    </li>
    <li class="page-item">
      <button class="page-link">3</button>
    </li>
    <li class="page-item">
      <div class="dropdown dropdown-root" id="overlay10">
        <button class="button-root btn btn-link nav-link dropdown-toggle page-link" type="button">
          <span class="text-truncate">... </span>
        </button>
      </div>
    </li>
    <li class="page-item">
      <button class="page-link">10</button>
    </li>
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
          <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
        </svg>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`PaginationBar should render with buttons if passed an onPageChange handler 1`] = `
<nav class="navbar pagination-bar-root">
  <div class="dropdown dropdown-root pagination-items-per-page" id="overlay3">
    <button class="button-root btn btn-secondary btn-sm dropdown-toggle" type="button">
      <span class="text-truncate">10 </span>
      <span>
        <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
        </svg>
      </span>
    </button>
  </div>
  <div class="pagination-results">showing-x-to-x-of-x-entries</div>
  <ul class="pagination pagination-root">
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
          <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
        </svg>
      </button>
    </li>
    <li class="page-item">
      <button class="page-link">1</button>
    </li>
    <li class="page-item">
      <button class="page-link">2</button>
    </li>
    <li class="page-item active">
      <button class="page-link">3</button>
    </li>
    <li class="page-item">
      <button class="page-link">4</button>
    </li>
    <li class="page-item">
      <button class="page-link">5</button>
    </li>
    <li class="page-item">
      <div class="dropdown dropdown-root" id="overlay4">
        <button class="button-root btn btn-link nav-link dropdown-toggle page-link" type="button">
          <span class="text-truncate">... </span>
        </button>
      </div>
    </li>
    <li class="page-item">
      <button class="page-link">10</button>
    </li>
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
          <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
        </svg>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`PaginationBar should render with large size 1`] = `
<nav class="navbar pagination-bar-root pagination-lg">
  <div class="dropdown dropdown-root pagination-items-per-page" id="overlay7">
    <button class="button-root btn btn-secondary btn-sm dropdown-toggle" type="button">
      <span class="text-truncate">10 </span>
      <span>
        <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
        </svg>
      </span>
    </button>
  </div>
  <div class="pagination-results">showing-x-to-x-of-x-entries</div>
  <ul class="pagination pagination-root">
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
          <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
        </svg>
      </button>
    </li>
    <li class="page-item">
      <button class="page-link">1</button>
    </li>
    <li class="page-item">
      <button class="page-link">2</button>
    </li>
    <li class="page-item active">
      <button class="page-link">3</button>
    </li>
    <li class="page-item">
      <button class="page-link">4</button>
    </li>
    <li class="page-item">
      <button class="page-link">5</button>
    </li>
    <li class="page-item">
      <div class="dropdown dropdown-root" id="overlay8">
        <button class="button-root btn btn-link nav-link dropdown-toggle page-link" type="button">
          <span class="text-truncate">... </span>
        </button>
      </div>
    </li>
    <li class="page-item">
      <button class="page-link">10</button>
    </li>
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
          <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
        </svg>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`PaginationBar should render with small size 1`] = `
<nav class="navbar pagination-bar-root pagination-sm">
  <div class="dropdown dropdown-root pagination-items-per-page" id="overlay5">
    <button class="button-root btn btn-secondary btn-sm dropdown-toggle" type="button">
      <span class="text-truncate">10 </span>
      <span>
        <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
        </svg>
      </span>
    </button>
  </div>
  <div class="pagination-results">showing-x-to-x-of-x-entries</div>
  <ul class="pagination pagination-root">
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
          <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
        </svg>
      </button>
    </li>
    <li class="page-item">
      <button class="page-link">1</button>
    </li>
    <li class="page-item">
      <button class="page-link">2</button>
    </li>
    <li class="page-item active">
      <button class="page-link">3</button>
    </li>
    <li class="page-item">
      <button class="page-link">4</button>
    </li>
    <li class="page-item">
      <button class="page-link">5</button>
    </li>
    <li class="page-item">
      <div class="dropdown dropdown-root" id="overlay6">
        <button class="button-root btn btn-link nav-link dropdown-toggle page-link" type="button">
          <span class="text-truncate">... </span>
        </button>
      </div>
    </li>
    <li class="page-item">
      <button class="page-link">10</button>
    </li>
    <li class="page-item">
      <button class="page-link">
        <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
          <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
        </svg>
      </button>
    </li>
  </ul>
</nav>
`;
